rpcs3/Emu/Cell/lv2/sys_fs.cpp

```diff
@@ -1653,15 +1666,83 @@ error_code sys_fs_fcntl(ppu_thread& ppu, u32 fd, u32 op, vm::ptr<void> _arg, u32
 
 	case 0xc0000007: // cellFsArcadeHddSerialNumber
 	{
-		const auto arg = vm::static_ptr_cast<lv2_file_c000007>(_arg);
+		const auto arg = vm::static_ptr_cast<lv2_file_c0000007>(_arg);
 		// TODO populate arg-> unk1+2
 		arg->out_code = CELL_OK;
 		return CELL_OK;
 	}
 
 	case 0xc0000008: // cellFsSetDefaultContainer, cellFsSetIoBuffer, cellFsSetIoBufferFromDefaultContainer
 	{
-		break;
+		// Allocates memory from a container/default container to a specific fd or default IO processing
+		const auto arg          = vm::static_ptr_cast<lv2_file_c0000008>(_arg);
+		auto& default_container = g_fxo->get<default_sys_fs_container>();
+
+		std::lock_guard def_container_lock(default_container.mutex);
+
+		if (fd == 0xFFFFFFFF)
+		{
+			// No check on container is done when setting default container
+			default_container.id   = arg->size ? ::narrow<u32>(arg->container_id) : 0u;
+			default_container.cap  = arg->size;
+			default_container.used = 0;
+
+			arg->out_code = CELL_OK;
+			return CELL_OK;
+		}
+
+		auto file = idm::get<lv2_fs_object, lv2_file>(fd);
+		if (!file)
+		{
+			return CELL_EBADF;
+		}
+
+		if (file->io_alloc_addr)
+		{
+			sys_memory_free(ppu, file->io_alloc_addr);
```

Comment:
It's not what I meant and altering user-memory area is a bad hack especially if coming from lv2.

Reply:
The changes done by taking and freeing memory should be observable as one by other threads, that's why I said adding the memory statd mutex lock.

Reply:
How is it a bad hack when it's what happens on ps3?
It's meant to allocate memory from a user container...

Reply:
LV2 memory has nothing to do with user memory.

Reply:
You said to use s_memstats_mtx but neither sys_memory_allocate_from_container nor sys_memory_free which basically do what we need to do(except for the allocating in user space) do.
From its use s_memstats_mtx seems there to protect global memory allocations(container creation and deletion as well as getting memory stats) so it doesn't seem to be what I should use.

I changed the take to be under idm lock and left the freeing under no lock(as the container can't be destroyed if allocated, which seems to be the logic under which sys_memory works).

Reply:
This is a multi-container operation - basically  transfering memory from one container to another. Creating and destroying containers are the same - multi-container operation only that the destination/source containervis the default container. That's why they all need to use that mutex.

Reply:
It doesn't transfer memory, this is a blocking operation if memory is in use. It deallocates the memory once it's not in use and the allocates memory on the new container/default. It's clearly not at the same time.

And it makes sense to have the mutex to take from the global container as it's not behind the idm lock.
I'm sorry not touching the current implementation without a good reason and so far I don't see one.

Reply:
This is just 3 lines of code, why all this pointless discussion about its atomicity nature.

Reply:
What happens if take fails? usually LV2 reverts all changes upon failure.

Reply:
Can you test it?

Reply:
I did, that's how I know.

Reply:
Okay, than it's probably not atomic if it does it this way.